crates/task-impls/src/consensus/helpers.rs

```diff
@@ -717,140 +717,215 @@ pub(crate) async fn handle_quorum_proposal_recv<TYPES: NodeType, I: NodeImplemen
+pub async fn decide_from_proposal<TYPES: NodeType>(
+    proposal: &QuorumProposal<TYPES>,
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+    existing_upgrade_cert: &Option<UpgradeCertificate<TYPES>>,
+    public_key: &TYPES::SignatureKey,
+) -> LeafChainTraversalOutcome<TYPES> {
+    let consensus_reader = consensus.read().await;
+    let view_number = proposal.view_number();
+    let parent_view_number = proposal.justify_qc.view_number();
+    let old_anchor_view = consensus_reader.last_decided_view();
 
-    #[allow(unused_mut)]
-    #[allow(unused_variables)]
-    let mut decided_upgrade_cert: Option<UpgradeCertificate<TYPES>> = None;
-    let mut new_anchor_view = consensus.last_decided_view();
-    let mut new_locked_view = consensus.locked_view();
-    let mut last_view_number_visited = view;
-    let mut new_commit_reached: bool = false;
-    let mut new_decide_reached = false;
-    let mut new_decide_qc = None;
-    let mut leaf_views = Vec::new();
-    let mut leafs_decided = Vec::new();
-    let mut included_txns = HashSet::new();
-    let old_anchor_view = consensus.last_decided_view();
-    let parent_view = proposal.justify_qc.view_number();
+    let mut last_view_number_visited = view_number;
     let mut current_chain_length = 0usize;
-    if parent_view + 1 == view {
-        current_chain_length += 1;
-        if let Err(e) = consensus.visit_leaf_ancestors(
-            parent_view,
-            Terminator::Exclusive(old_anchor_view),
-            true,
-            |leaf, state, delta| {
-                if !new_decide_reached {
-                    if last_view_number_visited == leaf.view_number() + 1 {
-                        last_view_number_visited = leaf.view_number();
-                        current_chain_length += 1;
-                        if current_chain_length == 2 {
-                            new_locked_view = leaf.view_number();
-                            new_commit_reached = true;
-                            // The next leaf in the chain, if there is one, is decided, so this
-                            // leaf's justify_qc would become the QC for the decided chain.
-                            new_decide_qc = Some(leaf.justify_qc().clone());
-                        } else if current_chain_length == 3 {
-                            new_anchor_view = leaf.view_number();
-                            new_decide_reached = true;
-                        }
-                    } else {
-                        // nothing more to do here... we don't have a new chain extension
-                        return false;
+    let mut res = LeafChainTraversalOutcome::default();
+
+    if let Err(e) = consensus_reader.visit_leaf_ancestors(
+        parent_view_number,
+        Terminator::Exclusive(old_anchor_view),
+        true,
+        |leaf, state, delta| {
+            let new_decide_reached = res.new_decided_view_number.is_some();
+
+            if !new_decide_reached {
```

Comment:
Gonna swap this and remove the above var, just waitng for CI before I push again.